addons/account/models/account_code_mapping.py

```diff
@@ -43,26 +44,45 @@ def _search(self, domain, offset=0, limit=None, order=None) -> Query:
-                existing_code_mappings = self.env.cache.get_records(self, self._fields['account_id']).filtered(
-                    lambda m: m.account_id.id in account_ids and m.company_id in companies
-                )
-                keys = {(m.account_id.id, m.company_id.id) for m in existing_code_mappings}
-                missing_keys = [
-                    (account_id, company_id)
-                    for account_id in account_ids
-                    for company_id in companies.ids
-                    if (account_id, company_id) not in keys
-                ]
-                max_existing_id = max(existing_code_mappings.ids, default=0)
-                new_code_mappings = self.browse(range(max_existing_id + 1, max_existing_id + len(missing_keys) + 1))
+                companies = companies.sorted(lambda c: (c.sequence, c.name))
 
-                self.env.cache.update(new_code_mappings, self._fields['account_id'], [account_id for account_id, _ in missing_keys])
-                self.env.cache.update(new_code_mappings, self._fields['company_id'], [company_id for _, company_id in missing_keys])
+                records = self._get_records(
+                    (account_id, company.id)
+                    for account_id in account_ids
+                    for company in companies
+                )
+                return records._as_query()
 
-                mappings = existing_code_mappings | new_code_mappings
-                return mappings.sorted(lambda m: (m.account_id.id, m.company_id.sequence, m.company_id.name))._as_query()
         raise NotImplementedError
 
+    @api.model
+    def _get_records(self, record_keys):
+        """ Get a recordset of code mappings for the given (account_id, company_id) record keys.
+
+            :param keys: an iterable of (account_id, company_id) record keys
+            :return: a recordset of code mappings corresponding to the record keys, in the same order.
+        """
+        record_keys = list(record_keys)
+        existing_records = self.env.cache.get_records(self, self._fields['account_id'])
+        record_id_by_key = {(record.account_id.id, record.company_id.id): record.id for record in existing_records}
+
+        missing_keys = [
+            record_key
+            for record_key in record_keys
+            if record_key not in record_id_by_key
+        ]
+
+        max_existing_id = max(existing_records.ids, default=0)
+        new_record_ids = range(max_existing_id + 1, max_existing_id + len(missing_keys) + 1)
+        new_records = self.browse(new_record_ids)
+
+        self.env.cache.update(new_records, self._fields['account_id'], [account_id for account_id, _ in missing_keys])
+        self.env.cache.update(new_records, self._fields['company_id'], [company_id for _, company_id in missing_keys])
+
+        record_id_by_key.update(zip(missing_keys, new_record_ids))
+
+        return self.browse(record_id_by_key[key] for key in record_keys)
```

Comment:
Can you remind me why we don't use the keys as id directly?

Reply:
We don't use the keys as id since in the front-end, the props validation forces resId to be a Number.
This means that a form view can't be opened for the record if its id is a string. 
see https://discord.com/channels/678381219515465750/687338039717920792/1288137412907630683

Reply:
That would be nice, but the problem with that is that it doesn't handle `account_id` being a `NewId`, which will be the case if we're editing the mapping for an account that's being created...